Integrations/src/main/java/io/deephaven/integrations/learn/Gatherer.java

```diff
@@ -0,0 +1,543 @@
+        final double[] tensor = new double[nRows * nCols];
+
+        try (final ResettableWritableDoubleChunk<? extends Values> valueChunk =
+                ResettableWritableDoubleChunk.makeResettableChunk();
+                final SharedContext sharedContext = SharedContext.makeSharedContext()) {
+
+            for (int ci = 0; ci < nCols; ++ci) {
+                valueChunk.resetFromArray(tensor, ci * nRows, nRows);
+                final ColumnSource<?> colSrc = columnSources[ci];
+
+                try (final ChunkSource.FillContext fillContext = colSrc.makeFillContext(nRows, sharedContext)) {
+                    // noinspection unchecked
+                    colSrc.fillChunk(fillContext, valueChunk, rowSeq);
+                }
+            }
+        }
+
+        return tensor;
+    }
+
+    /**
+     * Copy data from a table into a 2d tensor of doubles in row-major order.
+     *
+     * @param rowSeq indices of the rows of the table to put into the tensor
+     * @param columnSources columns of data to put into the tensor
+     * @return contiguous RAM allocated for the tensor. When a numpy tensor is passed in for this argument, jpy will
+     *         handle passing the memory reference as a 1d java array here.
+     */
+    public static double[] tensorBuffer2DDoubleRows(final RowSequence rowSeq, final ColumnSource<?>[] columnSources) {
+        final int nRows = rowSeq.intSize();
+        final int nCols = columnSources.length;
+        final double[] tensor = new double[nRows * nCols];
+
+        try (final SafeCloseableList toClose = new SafeCloseableList()) {
+            final RowSequence.Iterator rowKeys = toClose.add(rowSeq.getRowSequenceIterator());
+            final SharedContext inputSharedContext = toClose.add(SharedContext.makeSharedContext());
+            final ChunkSource.GetContext[] inputContexts = toClose.addArray(new ChunkSource.GetContext[nCols]);
+            for (int ci = 0; ci < nCols; ++ci) {
+                inputContexts[ci] = columnSources[ci].makeGetContext(COPY_CHUNK_SIZE, inputSharedContext);
+            }
+
+            // noinspection unchecked
+            final DoubleChunk<? extends Values>[] inputColumnValues = new DoubleChunk[nCols];
+            int ti = 0;
+            while (rowKeys.hasMore()) {
+                final RowSequence sliceRowKeys = rowKeys.getNextRowSequenceWithLength(COPY_CHUNK_SIZE);
+                for (int ci = 0; ci < nCols; ++ci) {
+                    inputColumnValues[ci] = columnSources[ci].getChunk(inputContexts[ci], sliceRowKeys).asDoubleChunk();
+                }
+
+                final int sliceChunkSize = sliceRowKeys.intSize();
+                for (int ri = 0; ri < sliceChunkSize; ++ri) {
+                    for (int ci = 0; ci < nCols; ++ci) {
+                        tensor[ti++] = inputColumnValues[ci].get(ri);
+                    }
+                }
+                inputSharedContext.reset();
+            }
+        }
+
+        return tensor;
+    }
+
```

Comment:
Just a note.  I'm redoing the primitives library.  This has some new code autogeneration stuff that will help you avoid having to hard-code these functions that only differ by types.  This is for future reference only.